### ISSUE TITLE
`unnamedResult`: false positive on functions returning aliased named types

### ISSUE THREAD

**Issue by hyeongyun0916** [go-critic/go-critic#1528]
## Summary

`unnamedResult` flags functions whose result list contains a Go 1.22+ type alias
(`type Foo = Bar`) as if the alias were an "unnamed" type, even when the alias
resolves to a perfectly named `*types.Named`. This produces false positives on
otherwise idiomatic code that uses aliases for re-export or import-cycle
avoidance.

```go
// Two distinct named types — should not be flagged.
type SomeAlias = somePkg.OtherType

func f() ([]uint32, *SomeAlias) { ... }
//   ^^^ false positive: "consider giving a name to these results"
```

## Root cause

`checkers/unnamedResult_checker.go`'s `typeName` helper has no `case *types.Alias`,
so an alias falls through to `default: return ""`. The rule then sees two
"nameless" types and concludes they collide, triggering the warning.

```go
func (c *unnamedResultChecker) typeName(typ types.Type) string {
    switch typ := typ.(type) {
    case *types.Array:    return c.typeName(typ.Elem())
    case *types.Pointer:  return c.typeName(typ.Elem())
    case *types.Slice:    return c.typeName(typ.Elem())
    case *types.Named:    return typ.Obj().Name()
    default:              return ""    // <-- *types.Alias lands here
    }
}
```

[`go/types.Alias`](https://pkg.go.dev/go/types#Alias) was introduced in Go 1.22.
With `GODEBUG=gotypesalias=1` (default since Go 1.23, mandatory once the flag
is removed) any `type Foo = Bar` declaration is represented by a `*types.Alias`
node distinct from the underlying `*types.Named`.

Notably, `sqlQuery_checker.go` in this same repository already handles
`*types.Alias` — so the team is aware of the type kind, it was just missed in
`unnamedResult_checker.go`.

## Relationship to #1526

[#1526](https://github.com/go-critic/go-critic/issues/1526) reports a similar
pattern (`typeName` returning `""` for unhandled kinds) but enumerates only
`*types.Basic`, `*types.Map`, and `*types.Chan` as the missing cases. It does
not mention `*types.Alias`. The fixes are orthogonal:

| Kind             | Triggered by                                         | Covered by |
|------------------|------------------------------------------------------|------------|
| `*types.Basic`   | `func() (int, string)` and similar                   | #1526      |
| `*types.Map`     | `func() (map[string]X, map[int]Y)`                   | #1526      |
| `*types.Chan`    | `func() (chan X, chan Y)`                            | #1526      |
| **`*types.Alias`** | `type Foo = Bar; func() (T, *Foo)`                 | **this issue** |

Even after #1526 lands, alias-using code still misfires because
`*types.Alias` is none of `Basic` / `Map` / `Chan`. The two need to be tracked
independently (or together in a broader fix — see "Proper fix" below).

## Reproduction

Append the following block to `checkers/testdata/unnamedResult/negative_tests.go`
(it relies on the existing `namedInt` / `namedStruct` definitions from
`positive_tests.go`):

```go
// --- type-alias regression tests ---
//
// Each function below uses a type alias (`type Foo = Bar`) in its result
// list. After Unalias-aware handling, every type resolves to a distinct
// named type, so the rule should be silent. With the current `typeName`
// switch (no `*types.Alias` case) the alias collapses to "" and the rule
// misfires.

type aliasedNamedInt = namedInt
type aliasedNamedStruct = namedStruct
type aliasedNamedIntChain = aliasedNamedInt

// 2-result list: pointer to alias of a named struct, distinct from `namedInt`.
func aliasedPointer() (namedInt, *aliasedNamedStruct) {
	return 0, nil
}

// 2-result list: two distinct aliases.
func aliasedDistinct() (aliasedNamedInt, aliasedNamedStruct) {
	return 0, aliasedNamedStruct{}
}

// 3-result list: two aliases of distinct named types + trailing error.
func aliasedTriple() (aliasedNamedInt, aliasedNamedStruct, error) {
	return 0, aliasedNamedStruct{}, nil
}

// Slices of two distinct aliased named types.
func aliasedSlice() ([]aliasedNamedInt, []aliasedNamedStruct) {
	return nil, nil
}

// Alias chain (alias of alias of named type).
func aliasedChain() (aliasedNamedIntChain, aliasedNamedStruct) {
	return 0, aliasedNamedStruct{}
}
```

Then run the suite:

```
$ go test ./checkers/ -run TestCheckers/unnamedResult
    linttest.go:171: testdata/unnamedResult/negative_tests.go:<L>: unexpected warn: consider giving a name to these results
    linttest.go:171: testdata/unnamedResult/negative_tests.go:<L>: unexpected warn: consider giving a name to these results
    linttest.go:171: testdata/unnamedResult/negative_tests.go:<L>: unexpected warn: consider giving a name to these results
    linttest.go:171: testdata/unnamedResult/negative_tests.go:<L>: unexpected warn: consider giving a name to these results
    linttest.go:171: testdata/unnamedResult/negative_tests.go:<L>: unexpected warn: consider giving a name to these results
--- FAIL: TestCheckers/unnamedResult
```

All five cases emit `unexpected warn`. Applying the one-line `*types.Alias`
case from the "Short fix" section below turns the failures into a clean pass.

## Short fix

Add one `*types.Alias` case that unwraps via the standard
[`types.Unalias`](https://pkg.go.dev/go/types#Unalias) helper:

```go
case *types.Alias:
    return c.typeName(types.Unalias(typ))
```

`Unalias` walks the alias chain to the first non-alias type, so chained aliases
(`type A = B; type C = A`) are handled in one shot. With this case in place,
the five repro functions stop firing.

**Happy to send the PR for this short fix** — minimal diff, no behavioral
change for code that doesn't use aliases, no API surface impact.

## Proper fix (open question for maintainers)

The "incomplete switch + silent `""` default" pattern is fragile. Whenever a
new `types.Type` kind appears (`Alias` in Go 1.22, more in future releases) or
an existing one is overlooked (`Basic` / `Map` / `Chan` per #1526), the
rule silently degrades into false positives.

A more durable fix would change the rule's signaling so "unable to name this
type" is distinguishable from "this type has no name", e.g.

- `typeName` returns `(string, ok bool)` and the rule skips the duplicate-name
  check when `ok` is false.
- Or use `types.Identical` for the "same type repeated" detection instead of
  comparing string names.
- Or recurse on `types.Unalias` at the entry point so the rule never sees an
  alias wrapper at all.

These are bigger changes — I'm happy to discuss but they'd need maintainer
direction on which way you'd prefer. Filing the short fix first as the
backwards-compatible minimum.

## Versions

- gocritic: `master` (verified on v0.13.0 and v0.14.3 — same bug in all).
- Go: 1.22+ (any version where `GODEBUG=gotypesalias=1`, which is the default
  since Go 1.23).

---

cc / related: #1526 (overlapping symptoms, disjoint scope)
